### PR TITLE
fix: circular dependencies part 1

### PR DESCRIPTION
**Pull request by nickolay-kondratyev**
# Pull Request Checklist

Part 1: of Circular Dependency Sequel.

In this part we start the split up/decoupling process. There isn't too many circular dependencies  knocked out (from 124 on master to 119) in this commit but it does the initial set up to start knocking out the circular dependencies. 

We will need to do this piece meal to try to minimize the merge conflicts as this sequel is going to be touching quite a lot of files. 

Using C# interface naming convention (to avoid attempting to rename current implementation). And file naming convention of suffixing `Interface.ts` to interface holding files to more align with our file naming convention. 

In this process, attempting to make no changes to the logic and move code around as much as possible in as is way. 

### Quality Assurance
- [x] Make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [x] Do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] If your tests involve running [[manual Testing|dendron://dendron.dendron-site/dendron.dev.qa.test#manual-testing]], make sure to update [[Test Workspace|dendron://dendron.dendron-site/dendron.dev.ref.test-workspace]] with any necessary notes/additions to perform manual test ^9jtWc6ov340p
- [ ] After you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: If you running MacOS or Linux, pay special attention to the Windows output and vice versa if you are developing on Windows

### CODE REVIEW

File: packages/plugin-core/src/ExtensionProvider.ts
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/ExtensionProvider.ts:5]: The idea with this class is that we can still query out statically to grab the extension (especially during transition) without having to depend on `workspace` file or concrete implementation of extension.

File: packages/plugin-core/src/_extension.ts
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/_extension.ts:981]: Commands are getting extension passed to their constructor to be able to use an instance of dependency injected extension interface rather than reaching out statically.

File: packages/plugin-core/src/commandFactory.ts
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/commandFactory.ts:21]: This allows commands that depend on instances of other commands to get instances of those commands without creating tight coupling to those implementations.

File: packages/plugin-core/src/commands/CreateNoteWithTraitCommand.ts
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/commands/CreateNoteWithTraitCommand.ts:144]: Since we already have getExtension() imported into these classes it doesn't make it worse to use it for now, but it also doesn't make it better as time goes on these should be refactored to use extension that gets passed into the instance of the command.

(or if something is used statically that we dont want to refactor right away then it should use ExtensionProvider to grab the interface of DendronExtension rather than coupling ourselves to concrete implementation of DendronExtension).

File: packages/plugin-core/src/dendronExtensionInterface.ts
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/dendronExtensionInterface.ts:31]: Going forward we should use IDendronExtension in places of concrete DendronExtension

**Comment by kevinslin** [comment]
Can you write out a high level description of what is being refactored here? will make the logic here a little easier to follow.

If there are best practices, we should add them to the `dev` section of [[Dev|dendron://dendron.docs/pkg.plugin-core.dev]]

File: packages/plugin-core/src/_extension.ts
Comment (jonathanyeung) [reply packages/plugin-core/src/_extension.ts:981]: A command with a no-arg constructor still works fine here right?

File: packages/plugin-core/src/commands/base.ts
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/base.ts:68]: hmm, was this a stray change?

File: packages/plugin-core/src/ExtensionProvider.ts
Comment (jonathanyeung) [reply packages/plugin-core/src/ExtensionProvider.ts:5]: Sounds good - can you add a comment to this class w/ this explanation? Also - make a note there that any _new_ classes should **not** use these statics and instead get an instance of `IDendronExtension` passed into their constructors

File: packages/plugin-core/src/dendronExtensionInterface.ts
Comment (jonathanyeung) [reply packages/plugin-core/src/dendronExtensionInterface.ts:31]: Can you go ahead and add some comments info for `IDendronExtension`? Doesn't need to be complete, we can build upon the docstrings over time.

File: packages/plugin-core/src/services/EngineAPIService.ts
Comment (jonathanyeung) [inline packages/plugin-core/src/services/EngineAPIService.ts:90]: nit: Why the removal of `public`? Was this done by an auto-code formatter? 

I understand that `public` is the default access level in typescript - but I personally like it there to explicitly indicate that it's meant for external consumption.

File: packages/plugin-core/src/workspace.ts
Comment (jonathanyeung) [inline packages/plugin-core/src/workspace.ts:574]: was it necessary to expose this one?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
Looking good

File: packages/plugin-core/src/services/EngineAPIService.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/services/EngineAPIService.ts:90]: It was an automatic refactor. 

Will add it back.

**Comment by jonathanyeung** [comment]
When we do check this in, would you mind running the circ dependency checker again, and adding the results to the bottom of [[Cyclical Dependency Fixing|dendron://private/scratch.2021.12.02.143546.cyclical-dependency-fixing]] (at least for plugin core circular deps)? So that we can track our progress and what's remaining.

**Comment by nickolay-kondratyev** [comment]
> Can you write out a high level description of what is being refactored here? will make the logic here a little easier to follow.

Added approach at the bottom of the task note [[dendron://private/task.workspace.2021.12.21.fix-blocking-circular-dependencies]]


> If there are best practices, we should add them to the `dev` section of [[Dev|dendron://dendron.docs/pkg.plugin-core.dev]]

Sounds good will do after we merge this first commit.

File: packages/plugin-core/src/workspace.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/workspace.ts:574]: No, put it back.

File: packages/plugin-core/src/commands/base.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/commands/base.ts:68]: This one was purposeful. 

We have commands that do not require options when they execute. Example `GoToNote`, So now with CommandFactory we are calling execute through the interface of BaseCommand. 

(As example we want to allow ShowPreview to call GoToNote command but we dont want ShowPreview to know the exact implementation of GoToNote).

File: packages/plugin-core/src/ExtensionProvider.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/ExtensionProvider.ts:5]: Added explanation, and @deprecated getExtension() in workspace.ts (along with getDWorkspace(),  getEngine()) with a comment to go their mirroring ExtensionProvider methods.

File: packages/plugin-core/src/_extension.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/_extension.ts:981]: Yep tested note lookup it has no-arg constructor and appears to work fine. 

```
NoteLookupCommand
  constructor() {
  ```

File: packages/plugin-core/src/dendronExtensionInterface.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/dendronExtensionInterface.ts:31]: Good point, added.

**Comment by nickolay-kondratyev** [comment]
Note commit that was blocked due to GoToNote circular dependency now passes : https://github.com/dendronhq/dendron/pull/1891

**Review by jonathanyeung — APPROVED** [review]

**Comment by jonathanyeung** [comment]
Looks good to me. @kevinslin - I'll let you review/approve prior to merging.

File: packages/plugin-core/src/commands/GotoNote.ts
Comment (kevinslin) [inline packages/plugin-core/src/commands/GotoNote.ts:65]: not required for this PR but does it make sense to make extension and wsUtils part of the command interface?

**Review by kevinslin — APPROVED** [review]
looks good, nice job!

File: packages/plugin-core/src/commands/GotoNote.ts
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/commands/GotoNote.ts:65]: By command interface do you mean adding IDendronExtension to BaseCommand?

**Pull request merged**
